### PR TITLE
Notification widget

### PR DESCRIPTION
**Pull request by patrickelectric**
Should be done:
- [x] Create a notification manager with everything needed to do so.
- [x] Create a singleton for icons, display information, style, colors.
  - This will helps both cpp and qml to access this kind of information without boilerplate. 
- [x] Move everything related to html requests to C++ as a service.
  - [x] Create a class to deal with updates and new versions.

### CODE REVIEW

File: src/network/networktool.cpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+#include <QUrl>
+
+PING_LOGGING_CATEGORY(NETWORKTOOL, "ping.networktool")
+
+void NetworkTool::checkInterfaceUpdate()
+{
+    // It's only necessary to get the gitUserRepo variable only once
+    static bool firstTime = true;
+    static QString gitUserRepo;
+    if(firstTime) {
```
Comment (Williangalvani) [inline src/network/networktool.cpp]: can't you just check if gitUserRepo is null instead?

File: src/network/networktool.cpp [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            gitUserRepo = regexMatch.capturedTexts()[1];
+        } else {
+            qCWarning(NETWORKTOOL) <<
+                                   "Fail to get github user and repository! It'll not be possible to check for updates";
+            return;
+        }
+        firstTime = false;
+    }
+
+    const QUrl url{QStringLiteral("https://api.github.com/repos/%1/releases").arg(gitUserRepo)};
```
Comment (Williangalvani) [inline src/network/networktool.cpp]: what happens if the first part failed and it gets here? fails silently?

File: src/notification/notificationmodel.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+bool NotificationModel::insertRows(int row, int count, const QModelIndex& parent)
+{
+
+    // Check for valid input
+    if (count < 1 || row < 0 || row > rowCount(parent)) {
+        return false;
+    }
+
+    // Add all request rows
+    while(count --> 0) {
```
Comment (Williangalvani) [inline src/notification/notificationmodel.cpp]:
```suggestion
    while(count-- > 0) {
```

File: qml/DisplaySettings.qml [hunk 4]
```diff
@@ -65,21 +67,15 @@ Item {
...
                         Layout.fillWidth: true
                         model: ["Dark", "Light"]
-                        onCurrentTextChanged: {
-                            if (currentIndex) {
-                                Style.useLightStyle()
-                            } else {
-                                Style.useDarkStyle()
-                            }
-                        }
+                        onCurrentTextChanged: StyleManager.isDark = !currentIndex
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply qml/DisplaySettings.qml]: Why this is wrong ?

File: qml/PingNotificationArea.qml [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            ScrollView {
+                id: scroll
+                clip: true
+                Layout.maximumHeight: window.height/2
+
+                ScrollBar.vertical: ScrollBar {
+                    // This will help to find which ScrollBar is not being used
+                    objectName: "verticalBar"
+                    interactive: false
+                    policy: ScrollBar.AlwaysOn
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply qml/PingNotificationArea.qml]: This helps the user to see if it exists new notifications without scrolling.

File: qml/PingNotificationArea.qml [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+                }
+
+                Component.onCompleted: {
+                    // This will delete all scrollbars that are not used
+                    // Doing: ScrollBar.vertical: ScrollBar {
+                    // does not remove old childrens
+                    // This is possible to check with gammaray
+                    // where it's possible to see the QQuickScrollBar items
+                    // that overlap right right and bottom part of the widget
+                    for(var i in scroll.data) {
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply qml/PingNotificationArea.qml]: This syntax does not exist in qml:
```qml
import QtQuick 2.7

Item {
    id: root

    Component.onCompleted: {
        var a = [3, 2, 1]
        for(var i of a) {
            print(i)
        }
    }
}

```

File: qml/PingNotificationArea.qml [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+                                    }
+
+
+                                    MouseArea {
+                                        anchors.fill: parent
+                                        // Do not handle clicks to expand text is a link
+                                        acceptedButtons: parent.hoveredLink ? Qt.NoButton : Qt.LeftButton
+                                        cursorShape: parent.hoveredLink ? Qt.PointingHandCursor : Qt.ArrowCursor
+                                        onClicked: {
+                                            if(!textItem.isBigText) {
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply qml/PingNotificationArea.qml]: This will make the indentation of the block bigger, and as consequence worst if the following lines get more complex.

File: src/network/networktool.cpp (hunk 2 above)
Comment (patrickelectric) [reply src/network/networktool.cpp]: What part ? Are you talking about `QStringLiteral` ?

File: src/notification/notificationmodel.cpp (hunk 3 above)
Comment (patrickelectric) [reply src/notification/notificationmodel.cpp]: I call it "big arrow operator".

**Comment by patrickelectric** [comment]
@Williangalvani 	corrections and comments done.

File: src/network/networktool.cpp (hunk 2 above)
Comment (Williangalvani) [reply src/network/networktool.cpp]: it still makes the request even with gitUserRepo empty?
nevermind, didn't notice the return there.

File: qml/DisplaySettings.qml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Williangalvani) [reply qml/DisplaySettings.qml]: well, first because you are processing the event as a single 'toggling',
then you are using boolean operators to toggle it between 1 and 0 ...

wouldn't work if you decided to add, for instance, a third colorblind mode
I don't think it is going to cause any trouble, though.

File: src/network/networktool.cpp (hunk 2 above)
Comment (Williangalvani) [reply src/network/networktool.cpp]: I meant it could try to make requests without gitUserRepo set. looks fine now.

File: qml/PingNotificationArea.qml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Williangalvani) [reply qml/PingNotificationArea.qml]: don't new ones show up on the top, pushing the old ones down?

File: src/notification/notificationmodel.cpp (hunk 3 above)
Comment (Williangalvani) [reply src/notification/notificationmodel.cpp]: nice

File: qml/PingNotificationArea.qml (hunk 5 above)
[the remark this answers is not in the document]
Comment (patrickelectric) [reply qml/PingNotificationArea.qml]: No, also this will make harder to read a selected event.

File: qml/PingNotificationArea.qml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Williangalvani) [reply qml/PingNotificationArea.qml]: sad

File: qml/Ping1DVisualizer.qml [hunk 8]
```diff
@@ -52,7 +53,7 @@ Item {
                 transform: Rotation { origin.x: height/2; angle: 90}
                 gradient: Gradient {
                     GradientStop { position: 0.3; color: "transparent" }
-                    GradientStop { position: 0.5; color: Style.primaryColor } // Not working with material
+                    GradientStop { position: 0.5; color: StyleManager.secondaryColor } // Not working with material
```
Comment (jaxxzer) [inline qml/Ping1DVisualizer.qml]: is the comment still relevant?

**Comment by jaxxzer** [comment]
Running doxygen gives `error: Project logo '../qml/imgs/ping.svg' specified by PROJECT_LOGO does not exist!`

File: qml/Ping1DVisualizer.qml (hunk 8 above)
Comment (patrickelectric) [reply qml/Ping1DVisualizer.qml]: Yep.

**Comment by patrickelectric** [comment]
@jaxxzer This PR do not changes ping.svg path, file or the Doxyfile. Besides that, try to run Doxygen inside the Doxyfile.

File: src/network/networkmanager.h [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+     */
+    static NetworkManager* self();
+
+    /**
+     * @brief Do json requests
+     * @param url URL to be requested
+     * @param function that will be called
+     *
+     */
+    static void requestJson(const QUrl& url, std::function<void(QJsonDocument&)> function);
```
Comment (jaxxzer) [inline src/network/networkmanager.h]: `function` -> `callback`?

File: qml/main.qml [hunk 10]
```diff
@@ -6,14 +6,18 @@ import QtGraphicalEffects 1.0
...
 ApplicationWindow {
     id: window
+    objectName: "window"
     title: "Ping Interface"
     visible: true
     height: 600
     width: 800
 
-    Material.theme: Style.theme
+    Material.theme: StyleManager.theme
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply qml/main.qml]: Material is already embedded in each controller, redoing that will be necessary to create each controller for our application just to not use Material.
 - https://github.com/qt/qtquickcontrols2/blob/5.11/src/imports/controls/material/material.pri

File: src/network/networkmanager.h (hunk 9 above)
Comment (patrickelectric) [reply src/network/networkmanager.h]: Yes, that's the idea. I'm using the function name because this variable is a std::function. A `callback`variable can be many things and a function is a function.

File: qml/Ping1DVisualizer.qml (hunk 8 above)
Comment (patrickelectric) [reply qml/Ping1DVisualizer.qml]: Because they are not working with Material. As the comment says.
Probably a bug in the signal propagation of Material, I'm not sure.

**Comment by jaxxzer** [comment]
Generally looks ok on first pass.
I am not sure about the NetworkManager and NetworkTool names, separation of roles and interfaces, but we can mature them more as needed.

File: qml/main.qml (hunk 10 above)
[the remark this answers is not in the document]
Comment (jaxxzer) [reply qml/main.qml]: ok!

File: src/network/networktool.cpp [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+#include <QUrl>
+
+PING_LOGGING_CATEGORY(NETWORKTOOL, "ping.networktool")
+
+void NetworkTool::checkInterfaceUpdate()
+{
+    static QString gitUserRepo;
+    if(gitUserRepo.isEmpty()) {
+        //https://github.com/user/repo.git
+        const static QRegularExpression regex("github.com/(.*).git");
```
Comment (jaxxzer) [inline src/network/networktool.cpp]:
```suggestion
        const static QRegularExpression regex("github.com/(.*)");
```
I don't put `.git`, it's not needed here or there.

File: src/network/networktool.cpp [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+#include <QTimer>
+#include <QUrl>
+
+PING_LOGGING_CATEGORY(NETWORKTOOL, "ping.networktool")
+
+void NetworkTool::checkInterfaceUpdate()
+{
+    static QString gitUserRepo;
+    if(gitUserRepo.isEmpty()) {
+        //https://github.com/user/repo.git
```
Comment (jaxxzer) [inline src/network/networktool.cpp]:
```suggestion
        //https://github.com/user/repo
```

File: src/network/networktool.cpp [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+#include <QJsonDocument>
+#include <QRegularExpression>
+#include <QTimer>
+#include <QUrl>
+
+PING_LOGGING_CATEGORY(NETWORKTOOL, "ping.networktool")
+
+void NetworkTool::checkInterfaceUpdate()
+{
+    static QString gitUserRepo;
```
Comment (jaxxzer) [inline src/network/networktool.cpp]:
```suggestion
    // check if the remote named `origin` is a github repository (determined at qmake-time) so we may use
    // the github api to check for available versions
    static QString gitUserRepo;
```

File: src/network/networktool.cpp [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+    );
+}
+
+void NetworkTool::checkNewVersionInGitHubPayload(QJsonDocument& jsonDocument)
+{
+    const static QString projectTag = QStringLiteral(GIT_TAG);
+    static QRegularExpression releaseTagRegex;
+    // Check if version start with 't' from test version
+    // For everything we can presume normal release
+    if(projectTag.contains('t') == 0) {
```
Comment (jaxxzer) [inline src/network/networktool.cpp]: // Check if version starts with 't' for test version
// Check if version starts with 'v' for release version

File: src/network/networktool.cpp [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+        QJsonValue json;
+    } lastReleaseAvailable;
+    lastReleaseAvailable.version = actualVersion;
+
+    // Check payload
+    for(const QJsonValue& versionPayload : versionsAvailable) {
+        auto versionString = versionPayload[QStringLiteral("tag_name")].toString();
+        qCDebug(NETWORKTOOL) << "Testing version:" << versionString;
+
+        regexMatch = releaseTagRegex.match(versionString);
```
Comment (jaxxzer) [inline src/network/networktool.cpp]: If we are currently running a test release, we will only be checking for test releases upstream
If we are currently running a regular release, we will only be checking for regular releases upstream (see above lines 46 and 49). Is this ok?

File: src/network/networktool.cpp [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+        if(!regexMatch.hasMatch()) {
+            continue;
+        }
+        qCDebug(NETWORKTOOL) << "Possible new version.";
+
+        auto version = versionString.right(1).toFloat();
```
Comment (jaxxzer) [inline src/network/networktool.cpp]: I think we need to talk about our versioning standards organization-wide. I like the simplicity of being able to simply parse a float, but I have a preference for Major.Minor.Patch semantic versioning.
https://semver.org/

File: src/network/networktool.cpp (hunk 15 above)
Comment (patrickelectric) [reply src/network/networktool.cpp]: Yes, that's the idea. Anyone that want to be bleeding edge or want to be a beta testerof ping-viewer will receive notifications about new test releases (a.k.a Beta).

File: src/network/networktool.cpp (hunk 16 above)
Comment (patrickelectric) [reply src/network/networktool.cpp]: In my mind, it depends of what kind of release cycle we want to use in the project.
For APIs, semver makes sense. For HUGE projects cmsver. And for simple projects calver or this one that I'm using, that is just new_versions.patches;

File: src/network/networktool.cpp (hunk 11 above)
Comment (jaxxzer) [reply src/network/networktool.cpp]: the regex will capture both cases with my suggested edit.

File: src/network/networktool.cpp (hunk 15 above)
Comment (jaxxzer) [reply src/network/networktool.cpp]: But these same people will not catch the stable release, even when it is newer than their beta/test.

File: src/network/networktool.cpp (hunk 16 above)
Comment (jaxxzer) [reply src/network/networktool.cpp]: @Williangalvani what is your thought?

File: src/notification/notificationmodel.cpp (hunk 3 above)
Comment (jaxxzer) [reply src/notification/notificationmodel.cpp]: Do you seriously like this @patrickelectric? The whitespace here looks insane to me. I'm surprised it is even legal.

File: src/notification/notificationmodel.h [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+    Q_INVOKABLE void remove(int index);
+
+    /**
+     * @brief Remove all items
+     *
+     */
+    Q_INVOKABLE void removeAll();
+
+private:
+    Q_DISABLE_COPY(NotificationModel)
```
Comment (jaxxzer) [inline src/notification/notificationmodel.h]: Why is this necessary?

Comment (jaxxzer) [reply src/notification/notificationmodel.h]: http://doc.qt.io/qt-5/qobject.html#Q_DISABLE_COPY

I see now, but why don't we do this on NotificationManager or our other QObject derived classes?

File: src/notification/notificationmodel.cpp [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+        for(auto* vector : _rows) {
```
Comment (jaxxzer) [inline src/notification/notificationmodel.cpp]: why do we use `auto*` here instead of just `auto`?

File: src/network/networktool.cpp (hunk 11 above)
Comment (patrickelectric) [reply src/network/networktool.cpp]: The problem is that "https://api.github.com/repos/bluerobotics/ping-viewer.git/releases" does not exist.

File: src/notification/notificationmodel.cpp (hunk 18 above)
Comment (patrickelectric) [reply src/notification/notificationmodel.cpp]: To not get a copy of it, AFAIK the compiler will set auto to T.

File: src/notification/notificationmodel.h (hunk 17 above)
Comment (patrickelectric) [reply src/notification/notificationmodel.h]: NotificationManager is a singleton, it's not possible to copy by it definition.
But about the other QObjects that we have yes. This line is a result based on Qt source code study of models that already exist.

File: src/network/networktool.cpp (hunk 15 above)
Comment (patrickelectric) [reply src/network/networktool.cpp]: Done.

**Comment by patrickelectric** [comment]
@jaxxzer Check if you agree with the corrections, let me know if something else need to be changed or merge as will.

File: src/notification/notificationmodel.h (hunk 17 above)
Comment (jaxxzer) [reply src/notification/notificationmodel.h]: The document that I linked says that any QObject should never be copied, not just singletons.

File: src/notification/notificationmodel.cpp (hunk 18 above)
Comment (jaxxzer) [reply src/notification/notificationmodel.cpp]: That's cool.

File: src/notification/notificationmodel.h (hunk 17 above)
Comment (patrickelectric) [reply src/notification/notificationmodel.h]: Check this line:
https://github.com/bluerobotics/ping-viewer/pull/315/files#diff-ef6adb74577178213a0ad3d437b8b25cR45

File: src/network/networktool.cpp (hunk 16 above)
Comment (jaxxzer) [reply src/network/networktool.cpp]: @patrickelectric I agree generally, but I think there is benefit to consistency in our organization or something we can add to our guidelines. Something that our users and engineers (current and future) can get used to. I agree that x.x.x may be more than we need here, but it's not _that_ excessive in my opinion and will be able to cover every project we do.

We don't need to reach decision before merge, but before we cut first stable.

ping @Williangalvani.

Comment (Williangalvani) [reply src/network/networktool.cpp]: I agree with @jaxxzer that consistency is important.
I understand that it is not required in this project right now, and a little more annoying to code with, but I vote for semver.
Or maybe we could have x.x standard for smaller projects?
Also, is your code assuming the version is only 1 digit?

File: src/notification/notificationmodel.h (hunk 17 above)
Comment (jaxxzer) [reply src/notification/notificationmodel.h]: hehe, it looks like this is the same effect and purpose of Q_DISABLE_COPY, so why not use that?

Comment (patrickelectric) [reply src/notification/notificationmodel.h]: Good point, I thought that was a different kind of disable copy but they are not. Will change it to make it uniform.
